### PR TITLE
User interface test branch 2

### PR DESCRIPTION
**Pull request by JayBSB**
User Interface and Permissions classes added according to the class diagram. Would appreciate a double check. :)

### CODE REVIEW

File: PMSCore/UserInterface.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (DistroRickDev) [inline PMSCore/UserInterface.cs]: Any particular reason for all this includes? Could we clean up this a bit?

File: PMSCore/Permission.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (DistroRickDev) [inline PMSCore/Permission.cs]: Maybe we could clean up this includes not sre they are all necessary

File: PMSCore/Permission.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    /// <summary>
+    /// Represents a user's permission with a level and description.
+    /// </summary>
+    public class Permission
+    {
+        /// <summary>
+        /// Gets the level of the permission.
+        /// A higher value indicates more access or privileges.
+        /// </summary>
+        public int PermissionLevel { get; private set; }
```
Comment (DistroRickDev) [inline PMSCore/Permission.cs]: maybe PermissionLevel could be of a type enum instead of an int what do you think?

File: PMSCore/Permission.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+        /// <summary>
+        /// Gets the description of the permission.
+        /// Provides additional context about what the permission level entails.
+        /// </summary>
+        public string PermissionDescription { get; private set; }
```
Comment (DistroRickDev) [inline PMSCore/Permission.cs]: private set?

File: PMSCore/Permission.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Permission"/> class with specific values.
+        /// </summary>
+        /// <param name="level">The permission level. Must be non-negative.</param>
+        /// <param name="description">The description of the permission.</param>
+        /// <exception cref="ArgumentException">Thrown if <paramref name="level"/> is negative.</exception>
+        public Permission(int level, string description)
+        {
+            if (level < 0)
+                throw new ArgumentException("Permission level must be a non-negative integer.", nameof(level));
```
Comment (DistroRickDev) [inline PMSCore/Permission.cs]: Careful with exception throwing this enforces the caller to try catch are we sure we wanna throw on object constructuion?

File: PMSCore/Permission.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            PermissionDescription = description;
+        }
+
+        /// <summary>
+        /// Returns a string representation of the permission.
+        /// </summary>
+        /// <returns>A string combining the permission level and description.</returns>
+        public override string ToString()
+        {
+            return $"Level {PermissionLevel}: {PermissionDescription}";
```
Comment (DistroRickDev) [inline PMSCore/Permission.cs]: If permission is an enum type it already implements this for you

**Review by DistroRickDev — COMMENTED** [review]
Looks good to me, very solid implementation, good test for happy path and corner cases, very well documented only left a few suggestion

**Comment by DistroRickDev** [comment]
On one last note I will only ask you to squash the commits into one

File: PMSCore/UserInterface.cs (hunk 1 above)
Comment (DistroRickDev) [reply PMSCore/UserInterface.cs]: Forgot to mention you can call IUser instead of UserInterface
https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/identifier-names

Comment (JayBSB) [reply PMSCore/UserInterface.cs]: Removed on latest push.

File: PMSCore/Permission.cs (hunk 2 above)
Comment (JayBSB) [reply PMSCore/Permission.cs]: Sorry, added by Visual Studio and forgot to remove. Removed.

File: PMSCore/Permission.cs (hunk 4 above)
Comment (JayBSB) [reply PMSCore/Permission.cs]: Changed implementation in newest update. I might be misunderstanding, but if we're using the permission as a DI shouldn't we have it as immutable outside of the constructor? Sorry if I did misunderstand. Lmk what you think when you see the updated implementation. :)

File: PMSCore/Permission.cs (hunk 5 above)
Comment (JayBSB) [reply PMSCore/Permission.cs]: Modified for updated code, exceptions now *very* basically handled. Thought I should probably update to just call the default constructor, but wasn't certain.

File: PMSCore/Permission.cs (hunk 6 above)
Comment (JayBSB) [reply PMSCore/Permission.cs]: Sorted :)

File: PMSCore/Permission.cs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                // Catch ArgumentOutOfRangeException and print to console.
+                // TODO: Update to use Logging.
+                Console.WriteLine(ex);
+            }
+            catch (Exception ex)
+            {
+                // Catch any other general exceptions and print to console.
+                // TODO: Update to use Logging.
+                Console.WriteLine(ex);
+            }
```
Comment (DistroRickDev) [inline PMSCore/Permission.cs]: https://learn.microsoft.com/en-us/dotnet/standard/exceptions/best-practices-for-exceptions

File: PMSCore/Permission.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        /// <summary>
+        public enum PermissionLevel
+        {
+            DEFAULT = 0,
+            USER = 1,
+            EMPLOYEE = 2,
+            MANAGER = 3,
+            ADMIN = 4,
+            SUPERUSER = 5
+        }
```
Comment (DistroRickDev) [inline PMSCore/Permission.cs]: This is all we need in this file is this enum no unit tests required, enums already come shiped with a serializer that allows to convert value to string by calling the ToString() method

**Comment by DistroRickDev** [comment]
At this point we need to simplify code here: IUser as pure interface, Permissions as an enum, in C# Enums are pretty packed up with helper functions, and no unit tests this task as it requires no implementation please adjust like wise if possible and squash the commits

**Comment by JayBSB** [comment]
> On one last note I will only ask you to squash the commits into one

I've been struggling pretty badly with that, but I am trying. Apologies for the delay.

EDIT: Figured it out!

**Comment by JayBSB** [comment]
> At this point we need to simplify code here: IUser as pure interface, Permissions as an enum, in C# Enums are pretty packed up with helper functions, and no unit tests this task as it requires no implementation please adjust like wise if possible and squash the commits

By this do you mean delete the method declarations in IUser, the descriptions from Permission and remove the tests for Permission as well?

**Comment by DistroRickDev** [comment]
> > At this point we need to simplify code here: IUser as pure interface, Permissions as an enum, in C# Enums are pretty packed up with helper functions, and no unit tests this task as it requires no implementation please adjust like wise if possible and squash the commits
> 
> By this do you mean delete the method declarations in IUser, the descriptions from Permission and remove the tests for Permission as well?

By this I mean just rename UserInterface to IUser and Permission should be just an enum not a class wrapping an enum (so Permissions dont need unit tests)

File: .editorconfig
Comment (DistroRickDev) [inline .editorconfig:1]: This file shouldn't be her

**Review by DistroRickDev — COMMENTED** [review]
Please rebase with latest main and remove .editorconfig if possible please

**Comment by JayBSB** [comment]
Branch abandoned due to git issues. Branch "user_interface" set up to apply changes.
